src/nvim/api/private/helpers.c

```diff
@@ -965,3 +966,41 @@ static void set_option_value_err(char *key,
     api_set_error(err, Exception, "%s", errmsg);
   }
 }
+
+
+/// Get an array containing dictionaries describing mappings
+/// based on mode and buffer id
+///
+/// @param  mode  The abbreviation for the mode
+/// @param  buf  The buffer to get the mapping array. NULL for global
+/// @returns An array of maparg() like dictionaries describing mappings
+ArrayOf(Dictionary) keymap_array(String mode, buf_T *buf)
+{
+  Array mappings = ARRAY_DICT_INIT;
+  dict_T *const dict = tv_dict_alloc();
+
+  // Convert the string mode to the integer mode
+  // that is stored within each mapblock
+  int int_mode = get_map_mode((char_u **)(&(mode.data)), 0);
```

Comment:
this would mutate `mode.data`, which is really unexpected, better use temporary `char_u *p = mode.data`.